alibi_detect/utils/pytorch/kernels.py

```diff
@@ -29,12 +36,76 @@ def sigma_median(x: torch.Tensor, y: torch.Tensor, dist: torch.Tensor) -> torch.
     return sigma
 
 
-class GaussianRBF(nn.Module):
+class BaseKernel(nn.Module):
+    """
+    The base class for all kernels.
+    """
+    def __init__(self) -> None:
+        super().__init__()
+        self.parameter_dict: dict = {}
+        self.active_dims: Optional[list] = None
+        self.feature_axis: int = -1
+
+    def forward(self, x: torch.Tensor, y: torch.Tensor, infer_parameter: bool = False) -> torch.Tensor:
+        raise NotImplementedError
+
+
+class SumKernel(nn.Module):
+    """
+    Construct a kernel by averaging two kernels.
+
+    Parameters:
+    ----------
+        kernel_a
+            the first kernel to be summed.
+        kernel_b
+            the second kernel to be summed.
+    """
     def __init__(
         self,
-        sigma: Optional[torch.Tensor] = None,
-        init_sigma_fn: Callable = sigma_median,
-        trainable: bool = False
+        kernel_a: BaseKernel,
+        kernel_b: BaseKernel
+    ) -> None:
+        super().__init__()
+        self.kernel_a = kernel_a
+        self.kernel_b = kernel_b
+
+    def forward(self, x: torch.Tensor, y: torch.Tensor,  infer_parameter: bool = False) -> torch.Tensor:
+        return (self.kernel_a(x, y, infer_parameter) + self.kernel_b(x, y, infer_parameter)) / 2
+
+
+class ProductKernel(nn.Module):
+    """
+    Construct a kernel by multiplying two kernels.
+
+    Parameters:
+    ----------
+        kernel_a
+            the first kernel to be summed.
+        kernel_b
+            the second kernel to be summed.
+    """
+    def __init__(
+        self,
+        kernel_a: BaseKernel,
+        kernel_b: BaseKernel
+    ) -> None:
+        super().__init__()
+        self.kernel_a = kernel_a
+        self.kernel_b = kernel_b
+
+    def forward(self, x: torch.Tensor, y: torch.Tensor,  infer_parameter: bool = False) -> torch.Tensor:
+        return self.kernel_a(x, y, infer_parameter) * self.kernel_b(x, y, infer_parameter)
+
+
+class GaussianRBF(BaseKernel):
+    def __init__(
+       self,
+       sigma: Optional[torch.Tensor] = None,
+       init_fn_sigma: Callable = sigma_median,
```

Comment:
Maybe we should have a standard kwarg name for this across all kernels. Perhaps now's a good time to change to something more generic like `param_initialiser`?

Reply:
Probs also a good time to achieve consistency by choosing only one of `infer` or `init`

Reply:
There is now an `infer_kernel_parameter` that does a loop over `parameter_dict`, so the overall behaviour is no longer dependent on the kernel.